pkg/controllers/addons_controller.go

```diff
@@ -39,224 +36,120 @@ import (
-		return
+	if l.IsPruningRequired() {
+		l.SetStatusPruning()
+		if err := l.Prune(); err != nil {
+			return err
+		}
+		l.SetDelayedRequeue()
+		return nil
 	}
-	// Start prune here in background thread with time limit of spec timeout value
-	// then wait for it to complete or timeout
-	time.Sleep(time.Second * 60)
-}
 
-func processPruning(l *layers.Layer, r *AddonsLayerReconciler) {
-	utils.Log(l.GetLogger(), 2, 1, "processing", "Status", kraanv1alpha1.PruningCondition)
-	// Check if pruning is done, if not restart prune here in background with time limit of spec timeout value
-	// Update status with details of pruning progress
-	// If completed, set status to ApplyPending
-}
+	if l.IsApplyRequired() {
+		if !l.DependenciesDeployed() {
+			l.SetDelayedRequeue()
+			return nil
+		}
 
-func processApplyPending(l *layers.Layer) {
-	utils.Log(l.GetLogger(), 2, 1, "processing", "Status", kraanv1alpha1.ApplyPendingCondition)
-	if !l.CheckK8sVersion() {
-		l.SetRequeue()
-		l.SetDelayed()
-		return
+		l.SetStatusApplying()
+		if err := l.Apply(); err != nil {
+			return err
+		}
+		l.SetDelayedRequeue()
+		return nil
 	}
-	// Wait for DependsOn layers to be Deployed
-}
 
-func processApply(l *layers.Layer, r *AddonsLayerReconciler) {
-	utils.Log(l.GetLogger(), 2, 1, "processing", "Status", kraanv1alpha1.ApplyCondition)
-	// Start apply here in background thread with time limit of spec timeout value
-	l.StatusApplying()
-}
-
-func processApplying(l *layers.Layer, r *AddonsLayerReconciler) {
-	utils.Log(l.GetLogger(), 2, 1, "processing", "Status", kraanv1alpha1.ApplyingCondition)
-	// Check if applying is done, if not restart applyhere in background with time limit of spec timeout value
-	// Update status with details of applying progress
-	// If completed, set status to Deployed
-}
-
-func processHold(l *layers.Layer) {
-	utils.Log(l.GetLogger(), 2, 0, "processing", "Status", kraanv1alpha1.HoldCondition)
-	if l.IsHold() {
-		l.SetHold()
+	if l.SuccessfullyApplied() {
+		l.SetStatusDeployed()
 	}
-	utils.Log(l.GetLogger(), 2, 1, "processed",
-		"Previous Status", kraanv1alpha1.HoldCondition,
-		"Status", l.GetFullStatus(),
-		"Spec:", l.GetSpec())
+	return nil
 }
 
-func processFailed(l *layers.Layer) {
-	utils.Log(l.GetLogger(), 2, 1, "processing", "Status", kraanv1alpha1.FailedCondition)
-	// Perform a retry if failed condition is more than 'interval' duration ago
-	// Use previous condition to detect what to retry
-	// However verify dependencies again in case something has changed in other AddonsLayers
-	// so effectively reset status to PrunePending or ApplyPending depending on what failed.
-	utils.Log(l.GetLogger(), 2, 1, "processed",
-		"Previous Status", kraanv1alpha1.FailedCondition,
-		"Status", l.GetFullStatus(),
-		"Spec:", l.GetSpec())
+func (r *AddonsLayerReconciler) updateRequeue(l layers.Layer, res *ctrl.Result, rerr *error) {
+	if l.IsUpdated() {
+		*rerr = r.update(r.Context, r.Log, l.GetAddonsLayer())
+	}
+	if l.NeedsRequeue() {
```

Comment:
I think we disagree on these being part of the `Layer`

Reply:
Your thinking r.NeedsRequeue(l Layer)?

Reply:
I'm thinking that the `Layer` shouldn't be aware of requeue at all :wink:

Reply:
Layer contains the implementation of the interface used to perform processing of AddonsLayer as part of reconcile.
In doing so it determines if the AddonsLayer is deployed or needs further processing and if a delay before reprocessing is appropriate. So it sets requeue and delay setting that the reconciler can use to decide whether to requeue now, requeue later or not to requeue. Happy to discuss this further on slack call if you want?

Reply:
Yeah lets chat